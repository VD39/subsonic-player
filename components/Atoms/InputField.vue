<script setup lang="ts">
withDefaults(
  defineProps<{
    error?: string;
    hideLabel?: boolean;
    id: string;
    label: string;
    modelValue: string | string[];
    placeholder?: string;
    required?: boolean;
    type?: string;
  }>(),
  {
    error: undefined,
    placeholder: undefined,
    type: 'text',
  },
);

const emit = defineEmits(['update:modelValue']);

function onInputInput(event: Event) {
  emit('update:modelValue', (event.target as HTMLInputElement).value);
}
</script>

<template>
  <div
    :class="[
      'column',
      $style.inputField,
      {
        [$style.error]: error,
      },
    ]"
  >
    <label
      ref="label"
      :class="[
        'mBXS',
        'sentenceCase',
        'smallFont',
        $style.label,
        {
          visuallyHidden: hideLabel,
        },
      ]"
      :for="id"
    >
      {{ label }}
      <span v-if="required" ref="required" :class="$style.required">*</span>
    </label>

    <input
      :id="id"
      ref="input"
      :type="type"
      autocomplete="off"
      :value="modelValue"
      :required="required"
      :placeholder="placeholder"
      :class="$style.input"
      @input="onInputInput"
    />

    <p v-if="error" ref="error" class="smallFont sentenceCase">
      {{ error }}
    </p>
  </div>
</template>

<style module>
.inputField {
  --label-color: var(--secondary-font-color);
  --input-border-color: var(--border-color);
  --input-width: var(--width-height-100);

  width: var(--input-width);
  color: var(--label-color);
}

.error {
  --label-color: var(--error-color);
  --input-border-color: var(--error-color);
}

.label {
  display: inline-block;
  color: var(--secondary-font-color);
}

.input {
  position: relative;
  width: var(--input-width);
  padding: calc(var(--space-12) - 2px) var(--space-12);
  color: var(--body-font-color);
  background-color: var(--secondary-background-color);
  border: 1px solid var(--input-border-color);
  border-radius: var(--border-radius-small);

  &::placeholder {
    opacity: 0.8;
  }
}

.required {
  color: var(--error-color);
}
</style>
