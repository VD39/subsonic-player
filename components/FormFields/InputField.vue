<script setup lang="ts">
withDefaults(
  defineProps<{
    errorMessage?: string;
    hasError?: boolean;
    hideLabel?: boolean;
    id: string;
    label: string;
    modelValue: string | string[];
    placeholder?: string;
    required?: boolean;
    type?: string;
  }>(),
  {
    errorMessage: undefined,
    placeholder: undefined,
    type: 'text',
  },
);

const emit = defineEmits(['update:modelValue']);

function updateModel(event: Event) {
  emit('update:modelValue', (event.target as HTMLInputElement).value);
}
</script>
<template>
  <div
    :class="[
      $style.inputField,
      {
        [$style.error]: hasError,
      },
    ]"
  >
    <label
      ref="label"
      :class="[
        $style.label,
        {
          'visually-hidden': hideLabel,
        },
      ]"
      :for="id"
    >
      {{ label }}
      <span v-if="required" ref="required" :class="$style.required">*</span>
    </label>

    <input
      :id="id"
      ref="input"
      :type="type"
      autocomplete="off"
      :value="modelValue"
      :required="required"
      :placeholder="placeholder"
      :class="$style.input"
      @input="updateModel"
    />

    <div v-if="hasError && errorMessage" ref="errorMessage">
      <p :class="$style.errorMessage">
        {{ errorMessage }}
      </p>
    </div>
  </div>
</template>

<style module>
.inputField {
  display: flex;
  flex-direction: column;
  width: 100%;
}

.label {
  display: inline-block;
  color: var(--secondary-font-color);

  &::first-letter {
    text-transform: capitalize;
  }
}

.input {
  position: relative;
  width: 100%;
  padding: calc(var(--space-12) - 1px) var(--space-12);
  color: var(--body-font-color);
  background-color: var(--secondary-background-color);
  border: 1px solid var(--border-color);
  border-radius: var(--border-radius-s);

  .error & {
    border-color: var(--error-color);

    &:focus {
      border-color: var(--error-color);
    }
  }

  &::placeholder {
    opacity: 0.5;
  }
}

.required {
  color: var(--error-color);
}

.errorMessage {
  margin-top: var(--space-2);
  font-size: var(--small-font-size);
  color: var(--error-color);

  &::first-letter {
    text-transform: capitalize;
  }
}
</style>
