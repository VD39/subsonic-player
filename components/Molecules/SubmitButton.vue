<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import SpinningLoader from '@/components/Atoms/SpinningLoader.vue';

const props = defineProps<{
  fullWidth?: boolean;
  loading?: boolean;
}>();

const icon = computed(() => (props.loading ? SpinningLoader : undefined));
</script>

<template>
  <ButtonLink
    :class="$style.submitButton"
    :disabled="loading"
    :fullWidth
    :icon
    iconPosition="right"
    showText
    type="submit"
  >
    <slot />
  </ButtonLink>
</template>

<style module>
.submitButton {
  --button-hover-color: var(--body-background-color);

  background-color: var(--button-hover-color);
  border: 1px solid var(--theme-color);
  border-radius: var(--border-radius-medium);

  @media (hover: hover) {
    &:hover,
    &:focus {
      --button-hover-color: var(--hover-selected-color);
    }
  }
}
</style>
