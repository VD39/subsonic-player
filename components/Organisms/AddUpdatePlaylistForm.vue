<script setup lang="ts">
import InputField from '@/components/Atoms/InputField.vue';
import SubmitButton from '@/components/Molecules/SubmitButton.vue';

const props = defineProps<{
  playlist?: Playlist;
}>();

const emit = defineEmits<{
  submit: [value: string | string[]];
}>();

const loading = ref(false);

const formInputs = {
  name: {
    validationRules: {
      required: true,
    },
    value: props.playlist?.name,
  },
};

const form = createForm(formInputs);

async function onFormSubmit() {
  validateInputs(form);

  if (!form.isValid.value) {
    return;
  }

  loading.value = true;

  emit('submit', form.fields.name.value.value);
}
</script>

<template>
  <form novalidate @submit.stop.prevent="onFormSubmit">
    <div class="formFields">
      <InputField
        :id="form.fields.name.id"
        v-model="form.fields.name.value.value"
        :error="form.fields.name.error.value"
        :label="form.fields.name.label"
        placeholder="Enter playlist name"
        :required="form.fields.name.required"
      />
    </div>

    <SubmitButton class="formButton" :loading>
      {{ playlist?.name ? 'Update' : 'Add' }} playlist
    </SubmitButton>
  </form>
</template>
