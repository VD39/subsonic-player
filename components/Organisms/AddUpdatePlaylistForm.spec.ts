import type { VueWrapper } from '@vue/test-utils';

import { mount } from '@vue/test-utils';

import InputField from '@/components/Atoms/InputField.vue';
import SubmitButton from '@/components/Molecules/SubmitButton.vue';
import { formattedPlaylistMock } from '@/test/fixtures';

import AddUpdatePlaylistForm from './AddUpdatePlaylistForm.vue';

function factory(props = {}) {
  return mount(AddUpdatePlaylistForm, {
    props: {
      ...props,
    },
  });
}

describe('AddUpdatePlaylistForm', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when playlist prop is not set', () => {
    it('sets the correct SubmitButton text', () => {
      expect(wrapper.findComponent(SubmitButton).text()).toBe('Add playlist');
    });
  });

  describe('when playlist prop is set', () => {
    beforeEach(() => {
      wrapper = factory({
        playlist: formattedPlaylistMock,
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('sets the correct SubmitButton text', () => {
      expect(wrapper.findComponent(SubmitButton).text()).toBe(
        'Update playlist',
      );
    });
  });

  describe('when form is invalid', () => {
    beforeEach(async () => {
      await wrapper.trigger('submit');
    });

    it('does not emit submit event', () => {
      expect(wrapper.emitted('submit')).toBe(undefined);
    });
  });

  describe('when form is valid', () => {
    beforeEach(async () => {
      wrapper
        .findComponent(InputField)
        .vm.$emit('update:modelValue', 'Playlist');
      await wrapper.trigger('submit');
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('emits submit event with the form values', () => {
      expect(wrapper.emitted('submit')).toEqual([['Playlist']]);
    });
  });
});
