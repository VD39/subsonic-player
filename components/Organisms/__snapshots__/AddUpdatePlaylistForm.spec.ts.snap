// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddUpdatePlaylistForm > matches the snapshot 1`] = `
"<form novalidate="">
  <div class="formFields">
    <div class="column inputField"><label class="mBXS sentenceCase smallFont label" for="name">name <span class="required">*</span></label><input id="name" type="text" autocomplete="off" required="" placeholder="Enter playlist name" class="input" value="">
      <!--v-if-->
    </div>
  </div><button class="centerItems buttonLink submitButton formButton submitButton formButton" type="submit">
    <!--v-if--><span class="clamp text">Add playlist </span>
  </button>
</form>"
`;

exports[`AddUpdatePlaylistForm > when form is valid > matches the snapshot 1`] = `
"<form novalidate="">
  <div class="formFields">
    <div class="column inputField"><label class="mBXS sentenceCase smallFont label" for="name">name <span class="required">*</span></label><input id="name" type="text" autocomplete="off" required="" placeholder="Enter playlist name" class="input" value="Playlist">
      <!--v-if-->
    </div>
  </div><button type="submit" class="centerItems buttonLink alignRight disabled submitButton formButton submitButton formButton" disabled=""><span class="centerItems spinningLoader icon" aria-hidden="true" weight="regular"><phcirclenotch size="24" class="icon"></phcirclenotch><span class="visuallyHidden">Loading...</span></span><span class="clamp text">Add playlist </span></button>
</form>"
`;

exports[`AddUpdatePlaylistForm > when playlist prop is set > matches the snapshot 1`] = `
"<form novalidate="">
  <div class="formFields">
    <div class="column inputField"><label class="mBXS sentenceCase smallFont label" for="name">name <span class="required">*</span></label><input id="name" type="text" autocomplete="off" required="" placeholder="Enter playlist name" class="input" value="name">
      <!--v-if-->
    </div>
  </div><button class="centerItems buttonLink submitButton formButton submitButton formButton" type="submit">
    <!--v-if--><span class="clamp text">Update playlist </span>
  </button>
</form>"
`;
